sdk/include/platform/sunburst/platform-ethernet.hh

```diff
@@ -0,0 +1,633 @@
+		               RxQueueControl::RxFrameCountThresholdEnable |
+		                 RxQueueControl::AutoDequeueRxQFrameEnable);
+
+		// Programmer's guide have a step to set the chip in half-duplex when
+		// negotiation failed, but we omit the step.
+
+		// Restart Port 1 auto-negotiation
+		register_set(RegisterOffset::P1CR, 1 << 13);
+
+		// Configure Low Watermark to 6KByte available buffer space out of
+		// 12KByte.
+		register_write(RegisterOffset::FCLWR, 0x0600);
+		// Configure High Watermark to 4KByte available buffer space out of
+		// 12KByte.
+		register_write(RegisterOffset::FCHWR, 0x0400);
+
+		// Clear the interrupt status
+		register_write(RegisterOffset::ISR, 0xFFFF);
+		receiveInterruptFutex =
+		  interrupt_futex_get(STATIC_SEALED_VALUE(EthernetInterruptCap));
+		// Enable Receive interrupt
+		register_write(RegisterOffset::IER, ReceiveInterrupt);
+
+		// Enable QMU Transmit.
+		register_set(RegisterOffset::TXCR, 1 << 0);
+		// Enable QMU Receive.
+		register_set(RegisterOffset::RXCR1, 1 << 0);
+	}
+
+	Ksz8851Ethernet(const Ksz8851Ethernet &) = delete;
+	Ksz8851Ethernet(Ksz8851Ethernet &&)      = delete;
+
+	/**
+	 * This device does not have a unique MAC address and so users must provide
+	 * a locally administered MAC address if more than one device is present on
+	 * the same network.
+	 */
+	static constexpr bool has_unique_mac_address()
+	{
+		return false;
+	}
+
+	static constexpr MACAddress mac_address_default()
+	{
+		return {0x3a, 0x30, 0x25, 0x24, 0xfe, 0x7a};
+	}
+
+	void mac_address_set(MACAddress address = mac_address_default())
+	{
+		register_write(RegisterOffset::MARH, (address[0] << 8) | address[1]);
+		register_write(RegisterOffset::MARM, (address[2] << 8) | address[3]);
+		register_write(RegisterOffset::MARL, (address[4] << 8) | address[5]);
+	}
+
+	uint32_t receive_interrupt_value()
+	{
+		return *receiveInterruptFutex;
+	}
+
+	int receive_interrupt_complete(Timeout *timeout,
+	                               uint32_t lastInterruptValue)
+	{
+		// If there are frames to process, do not enter wait.
+		if (frames_to_process)
+		{
+			return 0;
+		}
+
+		// Our interrupt is level-triggered; if a frame happens to arrive
+		// between `receive_frame` call and we marking interrupt as received,
+		// it will trigger again immediately after we acknowledge it.
+
+		// Acknowledge the interrupt in the scheduler.
+		interrupt_complete(STATIC_SEALED_VALUE(EthernetInterruptCap));
+		if (*receiveInterruptFutex == lastInterruptValue)
+		{
+			Debug::log("Acknowledged interrupt, sleeping on futex {}",
+			           receiveInterruptFutex);
+			return futex_timed_wait(
+			  timeout, receiveInterruptFutex, lastInterruptValue);
+		}
+		Debug::log("Scheduler announces interrupt has fired");
+		return 0;
+	}
+
+	/**
+	 * Simple class representing a received Ethernet frame.
+	 */
+	struct Frame
+	{
+		uint16_t length;
+		uint8_t *buffer;
+
+		Frame(uint16_t length) : length(length)
+		{
+			buffer = new uint8_t[length];
```

Comment:
new operator never returns null pointer -- it'll abort in OOM condition.

Reply:
I changed the code to do `new (std::nothrow)` and added nullptr checks.

Reply:
Not correct because we use the implementation with `std::nothrow_t`: https://github.com/microsoft/cheriot-rtos/blob/main/sdk/include/c%2B%2B-config/new

In any case I do not think that we should abort on an allocation error in the driver.

And we should maybe try to get rid of the allocation altogether and use a statically allocated buffer? This would make the error handling easier.

Reply:
(Note that my comment was left before I saw your latest comment)

Reply:
I'd prefer not have to implement a slab allocator inside the driver

Reply:
This doesn’t need a slab allocator, it needs a single static buffer large enough to hold one outgoing packet. For the Ethernet adaptor on the A7, this is some SRAM that’s owned by the device that the MAC can read from directly. For this, it can be a single buffer.

We should probably refactor this interface slightly so that adaptors like this need only to copy the packet headers and not the payload. We really only need the Ethernet + TCP/IP headers to be copied and checked, the rest of the packet can be copied from the source directly.